mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java

```diff
@@ -285,15 +240,71 @@ public MessageMetaData move(Mailbox mailbox, MailboxMessage original) throws Mai
     }
 
     @Override
-    protected MessageMetaData copy(Mailbox mailbox, MessageUid uid, long modSeq, MailboxMessage original)
+    public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws MailboxException {
+        messageMetadataMapper.enrichMessage(mailbox, message);
+
+        return save(mailbox, message);
+    }
+
+    @Override
+    public Iterator<UpdatedFlags> updateFlags(Mailbox mailbox, FlagsUpdateCalculator flagsUpdateCalculator,
+            MessageRange set) throws MailboxException {
+        ImmutableList.Builder<UpdatedFlags> updatedFlags = ImmutableList.builder();
+        Iterator<MailboxMessage> messages = findInMailbox(mailbox, set, FetchType.Metadata, -1);
+        
+        long modSeq = -1;
+        if (messages.hasNext()) {
+            modSeq = messageMetadataMapper.nextModSeq(mailbox);
+        }
+        while(messages.hasNext()) {
+            MailboxMessage member = messages.next();
+            Flags originalFlags = member.createFlags();
+            member.setFlags(flagsUpdateCalculator.buildNewFlags(originalFlags));
+            Flags newFlags = member.createFlags();
+            if (UpdatedFlags.flagsChanged(originalFlags, newFlags)) {
+                member.setModSeq(modSeq);
+                save(mailbox, member);
+            }
+
+            UpdatedFlags uFlags = new UpdatedFlags(member.getUid(), member.getModSeq(), originalFlags, newFlags);
+            
+            updatedFlags.add(uFlags);
+            
+        }
+
+        return updatedFlags.build().iterator();
+    }
+
+    @Override
+    public MessageMetaData copy(Mailbox mailbox, MailboxMessage original) throws MailboxException {
+        return copy(mailbox, messageMetadataMapper.nextUid(mailbox), messageMetadataMapper.nextModSeq(mailbox), original);  
+    }
+
+    @Override
+    public Optional<MessageUid> getLastUid(Mailbox mailbox) throws MailboxException {
+        return messageMetadataMapper.getLastUid(mailbox);
+    }
+
+    @Override
+    public long getHighestModSeq(Mailbox mailbox) throws MailboxException {
+        return messageMetadataMapper.getHighestModSeq(mailbox);
+    }
+
+    private MessageMetaData copy(Mailbox mailbox, MessageUid uid, long modSeq, MailboxMessage original)
             throws MailboxException {
         MailboxMessage copy;
+        JPAMailbox currentMailbox;
+        if (mailbox instanceof JPAMailbox) {
+            currentMailbox = (JPAMailbox) mailbox;
+        } else {
+            currentMailbox = new JPAMailbox(mailbox);
+        }
         if (original instanceof JPAStreamingMailboxMessage) {
```

Comment:
It depends on how can the DB can support, does it support streaming or encrypted data then system will support more advance feature (header and body can be streaming or can be encrypted). But for now, with guice, I've just seen default feature (no streaming and no encryption). So I don't know do we keep those or remove them all.